### PR TITLE
[`PEFT`] Fix `save_pretrained` to make sure adapters weights are also saved on TPU

### PR DESCRIPTION
**Pull request by shub-kris**
Bug Fix for saving adapter weights when using PEFT

# What does this PR do?

This PR fixes  saving adapter weights when using PEFT on TPUs. Currently only the model weights are being saved and not the adapter weights. 

I tested it locally with this change on this [script](https://gist.github.com/shub-kris/16fa07d7c98a4890d28a8366a203f7af) and now it saves following whiles whenever checkpointing:

```
README.md            
adapter_model.safetensors  
rng_state.pth  
special_tokens_map.json  
tokenizer.model        
trainer_state.json
adapter_config.json  
optimizer.pt               
scheduler.pt   
tokenizer.json           
tokenizer_config.json  
training_args.bin
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

Earlier discussed [here](https://huggingface.slack.com/archives/C01N44FJDHT/p1709227562328409)

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -3035,9 +3035,10 @@ def _save_tpu(self, output_dir: Optional[str] = None):
 
         # Save a trained model and configuration using `save_pretrained()`.
         # They can then be reloaded using `from_pretrained()`
+        supported_classes = (PreTrainedModel,) if not is_peft_available() else (PreTrainedModel, PeftModel)
```
Comment (ArthurZucker) [inline src/transformers/trainer.py:3038]: I think all `PushtoHubMixin` can fall in the classes that support save_pretrained and from pretrained so we could also use that as both should inherit from the latter

**Review by ArthurZucker — APPROVED** [review]
Thanks 🤗

**Comment by moficodes** [comment]
Hello there, What is the state of the PR? Is there a timeline when it will be merged and a new release of transformer will be out?

**Comment by ArthurZucker** [comment]
Just waiting for @shub-kris to come back ( he is off ) and transformers release will be in around 2 weeks

**Comment by moficodes** [comment]
I ran some tests on a GKE Cluster with TPU V4 with 4 nodes.

https://gist.github.com/moficodes/1492228c80a3c08747a973b519cc7cda

This run fails with 

```bash
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/safetensors/torch.py", line 13, in storage_ptr
    return tensor.untyped_storage().data_ptr()
RuntimeError: Attempted to access the data pointer on an invalid python storage.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "//fsdp.py", line 112, in <module>
    model.save_pretrained(new_model_id)
  File "/usr/local/lib/python3.10/site-packages/transformers/modeling_utils.py", line 2448, in save_pretrained
    safe_save_file(shard, os.path.join(save_directory, shard_file), metadata={"format": "pt"})
  File "/usr/local/lib/python3.10/site-packages/safetensors/torch.py", line 281, in save_file
    serialize_file(_flatten(tensors), filename, metadata=metadata)
  File "/usr/local/lib/python3.10/site-packages/safetensors/torch.py", line 470, in _flatten
    shared_pointers = _find_shared_tensors(tensors)
  File "/usr/local/lib/python3.10/site-packages/safetensors/torch.py", line 72, in _find_shared_tensors
    if v.device != torch.device("meta") and storage_ptr(v) != 0 and storage_size(v) != 0:
  File "/usr/local/lib/python3.10/site-packages/safetensors/torch.py", line 17, in storage_ptr
    return tensor.storage().data_ptr()
  File "/usr/local/lib/python3.10/site-packages/torch/storage.py", line 956, in data_ptr
    return self._data_ptr()
  File "/usr/local/lib/python3.10/site-packages/torch/storage.py", line 960, in _data_ptr
    return self._untyped_storage.data_ptr()
RuntimeError: Attempted to access the data pointer on an invalid python storage.
```

That looks like the original error. So not certain if the cause of the error was resolved.

**Comment by shub-kris** [comment]
> I ran some tests on a GKE Cluster with TPU V4 with 4 nodes.
> 
> https://gist.github.com/moficodes/1492228c80a3c08747a973b519cc7cda
> 
> This run fails with
> 
> ```shell
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.10/site-packages/safetensors/torch.py", line 13, in storage_ptr
>     return tensor.untyped_storage().data_ptr()
> RuntimeError: Attempted to access the data pointer on an invalid python storage.
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "//fsdp.py", line 112, in <module>
>     model.save_pretrained(new_model_id)
>   File "/usr/local/lib/python3.10/site-packages/transformers/modeling_utils.py", line 2448, in save_pretrained
>     safe_save_file(shard, os.path.join(save_directory, shard_file), metadata={"format": "pt"})
>   File "/usr/local/lib/python3.10/site-packages/safetensors/torch.py", line 281, in save_file
>     serialize_file(_flatten(tensors), filename, metadata=metadata)
>   File "/usr/local/lib/python3.10/site-packages/safetensors/torch.py", line 470, in _flatten
>     shared_pointers = _find_shared_tensors(tensors)
>   File "/usr/local/lib/python3.10/site-packages/safetensors/torch.py", line 72, in _find_shared_tensors
>     if v.device != torch.device("meta") and storage_ptr(v) != 0 and storage_size(v) != 0:
>   File "/usr/local/lib/python3.10/site-packages/safetensors/torch.py", line 17, in storage_ptr
>     return tensor.storage().data_ptr()
>   File "/usr/local/lib/python3.10/site-packages/torch/storage.py", line 956, in data_ptr
>     return self._data_ptr()
>   File "/usr/local/lib/python3.10/site-packages/torch/storage.py", line 960, in _data_ptr
>     return self._untyped_storage.data_ptr()
> RuntimeError: Attempted to access the data pointer on an invalid python storage.
> ```
> 
> That looks like the original error. So not certain if the cause of the error was resolved.

Hi @moficodes thanks for flagging this error, but on an initial glance, it doesn't look like the problem that this PR addresses. This PR aims to save the adapter weights, which were not being saved before this PR.

So, if you would have used trainer with this change it would save the adapter-weights too:

```
README.md            
adapter_model.safetensors  
rng_state.pth  
special_tokens_map.json  
tokenizer.model        
trainer_state.json
adapter_config.json  
optimizer.pt               
scheduler.pt   
tokenizer.json           
tokenizer_config.json  
training_args.bin
```
Earlier, `adapter_model.safetensors` and `adapter)config.json` were not being saved.

However, I will try to look into your error, reproduce it and get back to you.

File: src/transformers/trainer.py (hunk 1 above)
Comment (shub-kris) [reply src/transformers/trainer.py:3038]: @ArthurZucker perfect idea, let me give it a try :) and will update it.

Comment (shub-kris) [reply src/transformers/trainer.py:3038]: My fix was inspired from the code here: https://github.com/huggingface/transformers/blob/0cb946d3f2dd6c91dc9462ba3589d35e7bcbc618/src/transformers/trainer.py#L3073

Comment (shub-kris) [reply src/transformers/trainer.py:3038]: @ArthurZucker if I understood correctly, I tried the following change:

`supported_classes = (PushToHubMixin, )` in [line 3073](https://github.com/huggingface/transformers/blob/0cb946d3f2dd6c91dc9462ba3589d35e7bcbc618/src/transformers/trainer.py#L3073) but it didn't work.

**Comment by moficodes** [comment]
I see. Will open a separate issue for it then. 

Thank you!

**Comment by moficodes** [comment]
The error happens on the same line though.  `model.save_pretrained(new_model_id)`

**Comment by shub-kris** [comment]
@LysandreJik can we merge this if it looks good to you, since @ArthurZucker is on holidays and I made the changes he asked and have tested it too locally.

**Comment by shub-kris** [comment]
@moficodes answered it here: https://github.com/huggingface/transformers/issues/29608#issuecomment-1993797013

**Comment by amyeroberts** [comment]
@shub-kris Based on reviews and code, we can merge. There's currently a failing test which needs to be resolved first. Could you try rebasing on main to make sure you have all the latest updates, and trigger a fresh CI run?

**Comment by shub-kris** [comment]
@amyeroberts some checks are still failing because of this I guess: https://github.com/huggingface/transformers/runs/22627188153

**Comment by amyeroberts** [comment]
@shub-kris Yep - a fix has just been merged into main. Apologies for the disruption. Could you try rebasing again?

**Comment by shub-kris** [comment]
Thanks a lot @amyeroberts

**Pull request merged**